### ISSUE TITLE
Support for S3 repositories

### ISSUE THREAD

**Issue by bryandatanerds** [tadfisher/gradle2nix#32]
Hello,

I've been trying to get a previously working Nix build of a project going again after various upgrades of Gradle and Spring, and the current issue I'm facing is that we have an S3 bucket that we use as a Maven repository, but this seems not to work now when calling gradle2nix, perhaps because of the new dependency resolution introduced in 648be6bd071309705937d71ecf62b735d32f27c8.  The error message is below.

Is there an easy way to fix this so that gradle2nix (well, Ivy) is able to read from the repository, using Gradle's `awsIm(AwsImAuthentication)` or some other means?

If that's not easy, I'd also be interested in having a way to tell (or patch...) gradle2nix to ignore missing dependencies, since we already have custom logic in Nix to replace the gradle-env.json deps for things coming from this repo with locally built jars/poms instead.

Thanks!

    FAILURE: Build failed with an exception.

    * What went wrong:
    maven: unable to get resource for org/springframework/boot#spring-boot-configuration-processor;2.3.4.RELEASE: res=s3://mybucket/org/springframework/boot/spring-boot-configuration-processor/2.3.4.RELEASE/spring-boot-configuration-processor-2.3.4.RELEASE.pom: java.net.MalformedURLException: unknown protocol: s3
    > unknown protocol: s3

    * Try:
    Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

    * Get more help at https://help.gradle.org

    CONFIGURE FAILED in 47s
    Exception in thread "main" org.gradle.tooling.BuildException: Could not fetch model of type 'Build' using connection to Gradle distribution 'https://services.gradle.org/distributions/gradle-6.7-bin.zip'.
            at org.gradle.tooling.internal.consumer.ExceptionTransformer.transform(ExceptionTransformer.java:51)
            at org.gradle.tooling.internal.consumer.ExceptionTransformer.transform(ExceptionTransformer.java:29)
    <...much more stack trace...>
    Caused by: java.lang.RuntimeException: maven: unable to get resource for org/springframework/boot#spring-boot-configuration-processor;2.3.4.RELEASE: res=s3://mybucket/org/springframework/boot/spring-boot-configuration-processor/2.3.4.RELEASE/spring-boot-configuration-processor-2.3.4.RELEASE.pom: java.net.MalformedURLException: unknown protocol: s3
            at org.apache.ivy.plugins.resolver.RepositoryResolver.findResourceUsingPattern(RepositoryResolver.java:139)
            at org.apache.ivy.plugins.resolver.AbstractPatternsBasedResolver.findResourceUsingPatterns(AbstractPatternsBasedResolver.java:94)
            at org.apache.ivy.plugins.resolver.IBiblioResolver.findArtifactRef(IBiblioResolver.java:129)
            at org.apache.ivy.plugins.resolver.BasicResolver.getArtifactRef(BasicResolver.java:1029)
            at org.apache.ivy.plugins.resolver.BasicResolver.locate(BasicResolver.java:894)
            at org.apache.ivy.plugins.resolver.IBiblioResolver.locate(IBiblioResolver.java:575)
            at org.nixos.gradle2nix.MavenResolver.resolve(RepositoryResolver.kt:73)
            at org.nixos.gradle2nix.RepositoryResolver.resolve$default(RepositoryResolver.kt:52)
            at org.nixos.gradle2nix.ConfigurationResolver.resolvePoms(ConfigurationResolver.kt:142)
            at org.nixos.gradle2nix.ConfigurationResolver.resolveMetadata(ConfigurationResolver.kt:115)
            at org.nixos.gradle2nix.ConfigurationResolver.resolve(ConfigurationResolver.kt:83)
            at org.nixos.gradle2nix.Gradle2NixPluginKt.projectDependencies(Gradle2NixPlugin.kt:227)
            at org.nixos.gradle2nix.Gradle2NixPluginKt.buildProject(Gradle2NixPlugin.kt:184)

**Comment by tadfisher** [comment]
Yeah, this is an unfortunate side-effect of using an external dependency resolver like Ivy. While Gradle's internal resolver originated from Ivy, it has diverged enough that it is now a moving target for anyone trying to emulate it.

Unfortunately, Gradle's internal resolution code is sprawling and deliberately not well-exposed to the outside world. We need to use internal APIs just to obtain the the project's repositories, for instance, and simply obtaining a list of URLs fetched during dependency resolution is essentially impossible without reflection and version-specific hacks. This is what the external resolver was written to solve, as we need to resolve dependencies from *all* repositories in order to reproduce the single Maven namespace that is provided by the dependency model in Gradle.

To implement this in the external resolver, then, we need to:
- Get credentials for repositories in Gradle's `RepositoryHandler`.
- Hook into or wrap Ivy's `RepositoryResolver` to provide the mechanism for downloading artifacts from S3 with credentials support.
- Somehow test this; either find a free s3-compatible storage provider or, better, find something that works like jimfs but runs a s3-compatible server backed by the local filesystem or even in-memory data.

In the short term, adding support for per-dependency overrides is needed. I'll spec this out; is there a syntax that you would prefer?

**Comment by tadfisher** [comment]
@bryandatanerds I added preliminary support for resolving dependencies from S3 in the [s3 branch](https://github.com/tadfisher/gradle2nix/compare/s3). Would you mind trying this out?

If you are using Nix 2.4 with flakes support, you can run:

```
nix run github:tadfisher/gradle2nix/s3
```

Otherwise, you can build it using:

```
nix build -f "https://github.com/tadfisher/gradle2nix/archive/s3.tar.gz"
```

**Comment by tadfisher** [comment]
@bryandatanerds I'm going to go ahead and merge #35, which should address this. If the S3 support doesn't work for you, feel free to reopen this.

**Comment by bryandatanerds** [comment]
Hi @tadfisher, thanks for your replies to both of my issues and sorry for the radio silence.  I've been busy and haven't had a chance to come back and respond properly (plus I managed to work around both of the issues, so it was a case of being good enough for the moment).  I did run this and I got an exception while gradle2nix was running, so I'll try again and if the problem persists I'll post back.  Cheers.
